### ISSUE TITLE
🎯 Implementation Plan: ADE Pipeline with Full Dataset & Multi-Signature Optimization

### ISSUE THREAD

**Issue by vicentereig** [vicentereig/dspyrb-examples#6]
# 🚀 ADE Pipeline with Full Dataset & Optimization

### 1.1 DrugExtractor Signature
**Purpose**: Extract drug names from medical text
```ruby
class DrugExtractor < DSPy::Signature
  input :text, String  # Medical text
  output :drugs, T::Array[String]  # Extracted drug names
end
```

**KPIs:**
- **Precision**: Of all drugs extracted, how many are actually in the text?
  - *Medical context*: Avoid false positives (imaginary drugs)
- **Recall**: Of all drugs in the text, how many were found?
  - *Medical context*: Don't miss any medications
- **F1 Score**: Harmonic mean of precision and recall
  - *Medical context*: Balance between completeness and accuracy

### 1.2 EffectExtractor Signature
**Purpose**: Extract adverse effects from medical text
```ruby
class EffectExtractor < DSPy::Signature
  input :text, String  # Medical text
  output :effects, T::Array[String]  # Extracted adverse effects
end
```

**KPIs:**
- **Precision**: Of all effects extracted, how many are actual ADEs?
  - *Medical context*: Avoid normal symptoms being labeled as ADEs
- **Recall**: Of all ADEs in the text, how many were found?
  - *Medical context*: Critical for patient safety - don't miss ADEs
- **F1 Score**: Balance metric
  - *Medical context*: Optimize for high recall (safety first)

### 1.3 ADEClassifier Signature
**Purpose**: Binary classification of ADE presence
```ruby
class ADEClassifier < DSPy::Signature
  input :text, String
  input :drugs, T::Array[String]
  input :effects, T::Array[String]
  output :has_ade, T::Boolean  # true=ADE present, false=No ADE
  output :confidence, Float
end
```

**KPIs:**
- **Precision**: Of all ADE predictions, how many are correct?
  - *Medical context*: Avoid false alarms
- **Recall**: Of all actual ADEs, how many were detected?
  - *Medical context*: CRITICAL - missing ADEs can be fatal
- **F1 Score**: Overall performance
- **Specificity**: True negative rate
  - *Medical context*: Avoid unnecessary interventions

### 1.4 Full Pipeline Metrics
**End-to-end performance:**
- **Overall Accuracy**: Correct classifications / Total
- **Weighted F1**: Accounts for class imbalance (71% negative)
- **False Negative Rate**: Most critical metric for medical safety
- **Matthews Correlation Coefficient (MCC)**: Better for imbalanced data

### Directories to remove entirely:
```
changelog_generator/     # Not relevant to ADE task
fixtures/llm-changelog-generator/  # Changelog specific
reports/                 # Old changelog reports
optimization_results/    # Old results with wrong structure
```

### Files to delete:
```
# Changelog-related
run_batch_comparison.rb
run_monolithic_comparison.rb
compare_batch_vs_structured_output.rb
analyze_batch_token_usage.rb
BATCH_TOKEN_ANALYSIS.md

# Old dataset files
dataset_sample_for_blog.rb
dataset_sample_for_blog.json
data/0000.parquet  # Old single file

# Old optimization outputs
optimization_output*.txt
analyze_improvement_metrics.rb
improvement_metrics_analysis.json

# Test files not needed
test_empty_input.rb
test_structured_output.rb
demo_cost_tracking.rb
```

### Keep but refactor:
```
lib/ade_predictor.rb      # Convert to binary
lib/medical_text_extractor.rb  # Split into DrugExtractor + EffectExtractor
lib/dataset_loader.rb     # Use polars-df
lib/evaluation_metrics.rb # Simplify for binary
lib/baseline_predictor.rb # Update for new signatures
lib/ade_optimizer.rb      # Keep both optimizers
```

## 📁 3. New File Structure

```
dspyrb-examples/
├── data/
│   └── ade_corpus_v2/    # Keep downloaded parquet files
├── lib/
│   ├── signatures/
│   │   ├── drug_extractor.rb
│   │   ├── effect_extractor.rb
│   │   └── ade_classifier.rb
│   ├── pipeline/
│   │   ├── ade_pipeline.rb
│   │   └── baseline_pipeline.rb
│   ├── data/
│   │   └── ade_dataset_loader.rb  # Polars-based
│   ├── evaluation/
│   │   ├── extraction_metrics.rb
│   │   ├── classification_metrics.rb
│   │   └── pipeline_metrics.rb
│   └── optimization/
│       └── pipeline_optimizer.rb
├── scripts/
│   ├── run_baseline_evaluation.rb
│   ├── run_optimization.rb
│   └── compare_optimizers.rb
└── results/
    └── [timestamp]/
        ├── baseline_metrics.json
        ├── simple_optimizer_metrics.json
        └── miprov2_metrics.json
```

### Phase 1: Data Infrastructure
1. Create `AdeDatasetLoader` using polars-df
2. Load all 3 parquet files
3. Create unified dataset with:
   - 23,516 classification examples
   - 6,821 drug-effect annotations
   - 279 drug-dosage annotations
4. Split into train/val/test (70/15/15)

### Phase 2: Signatures
1. Implement `DrugExtractor` signature
2. Implement `EffectExtractor` signature
3. Implement `ADEClassifier` signature
4. Create type definitions for outputs

### Phase 3: Baseline Pipeline
1. Create unoptimized pipeline chaining all 3 signatures
2. Measure baseline performance on validation set
3. Calculate all KPIs per signature and overall

### Phase 4: Optimization
1. **SimpleOptimizer**:
   - Optimize each signature independently
   - Use drug_ade_relation data for extractors
   - Use classification data for classifier
   - Measure improvements

2. **MIPROv2**:
   - Same training data
   - Bootstrap optimization
   - Compare with SimpleOptimizer

### Phase 5: Evaluation & Comparison
1. Run both optimizers on test set
2. Generate comparison report:
   ```
   Metric         | Baseline | SimpleOpt | MIPROv2
   ------------------------------------------------
   Drug Recall    |   45%    |    68%    |   75%
   Effect Recall  |   40%    |    65%    |   72%
   ADE F1 Score   |   55%    |    71%    |   78%
   False Neg Rate |   35%    |    20%    |   15%
   ```

### Minimum Viable Performance:
- Drug extraction recall > 70%
- Effect extraction recall > 65%
- ADE classification F1 > 70%
- False negative rate < 20%

### Target Performance (with MIPROv2):
- Drug extraction recall > 80%
- Effect extraction recall > 75%
- ADE classification F1 > 80%
- False negative rate < 15%

### For DrugExtractor:
- Train on 6,821 drug annotations from drug_ade_relation
- Additional 279 from drug_dosage_relation
- Ground truth: exact drug names from dataset

### For EffectExtractor:
- Train on 6,821 effect annotations from drug_ade_relation
- Ground truth: exact adverse effects from dataset

### For ADEClassifier:
- Train on all 23,516 classification labels
- Input includes extracted drugs & effects from previous steps
- Handle class imbalance with weighted metrics

## 🔬 7. Experiment Tracking

Each optimization run will track:
1. **Per-signature metrics**: precision, recall, F1
2. **Pipeline metrics**: end-to-end accuracy
3. **Medical safety metrics**: false negative rate, missed ADEs
4. **Optimization metadata**: 
   - Optimizer type
   - Number of training examples
   - Number of demos (k)
   - Optimization iterations

## 💾 8. Output Format

```json
{
  "timestamp": "2024-08-12T10:00:00Z",
  "dataset": {
    "train": 16461,
    "val": 3527,
    "test": 3528
  },
  "baseline": {
    "drug_extractor": { "precision": 0.65, "recall": 0.45, "f1": 0.53 },
    "effect_extractor": { "precision": 0.60, "recall": 0.40, "f1": 0.48 },
    "ade_classifier": { "precision": 0.70, "recall": 0.50, "f1": 0.58 },
    "pipeline": { "accuracy": 0.65, "f1": 0.55, "false_neg_rate": 0.35 }
  },
  "simple_optimizer": {
    "improvements": { "f1": "+16%", "recall": "+23%" },
    "final_metrics": { ... }
  },
  "miprov2": {
    "improvements": { "f1": "+23%", "recall": "+32%" },
    "final_metrics": { ... }
  }
}
```

## 🎬 9. Next Steps

1. [ ] Clean up codebase - remove irrelevant files
2. [ ] Implement new signatures (DrugExtractor, EffectExtractor, ADEClassifier)
3. [ ] Create AdeDatasetLoader with polars-df
4. [ ] Build baseline pipeline
5. [ ] Run SimpleOptimizer optimization
6. [ ] Run MIPROv2 optimization
7. [ ] Compare results and generate report
8. [ ] Document findings for blog post

## 📝 Related Issues

- #5: ADE Dataset Structure & Polars Integration (context gathering for blog)
- #2: Achievement: 67% API call reduction with batch processing

This plan creates a clean, focused pipeline that:
1. Uses real medical data (not synthetic)
2. Optimizes three distinct signatures
3. Compares both available optimizers
4. Tracks medical-specific KPIs
5. Maintains reproducibility with VCR
